db/memorystore.go

```diff
@@ -437,3 +441,51 @@ func (m *MemoryStore) GetExtenalAccountKeyByID(keyID string) ([]byte, bool) {
 	key, ok := m.externalAccountKeysByID[keyID]
 	return key, ok
 }
+
+// AddBlockedDomain will add the domain name to the block list
+func (m *MemoryStore) AddBlockedDomain(name string) error {
+	if len(name) == 0 {
+		return errors.New("domain name must not be empty")
+	}
+
+	m.Lock()
+	defer m.Unlock()
+	domainParts := strings.Split(name, ".")
+
+	// reversing the order
+	for i, j := 0, len(domainParts)-1; i < j; i, j = i+1, j-1 {
+		domainParts[i], domainParts[j] = domainParts[j], domainParts[i]
+	}
+
+	m.blockListByDomain = append(m.blockListByDomain, domainParts)
+
+	return nil
+}
+
+// IsDomainBlocked will return true if a domain is on the block list
+func (m *MemoryStore) IsDomainBlocked(name string) bool {
+	m.RLock()
+	defer m.RUnlock()
+
+	domainParts := strings.Split(name, ".")
+
+	// reversing the order
+	for i, j := 0, len(domainParts)-1; i < j; i, j = i+1, j-1 {
+		domainParts[i], domainParts[j] = domainParts[j], domainParts[i]
+	}
+
+	for _, blockedParts := range m.blockListByDomain {
+		isMatch := true
+		for i := range blockedParts {
+			if blockedParts[i] != domainParts[i] {
+				isMatch = false
+				break
+			}
+		}
```

Comment:
An alternate approach here would be to store the blocked domains as strings, concatenate domainParts back into a single string, and then do a prefix check: https://golang.org/pkg/strings/#HasPrefix

I'm honestly not sure which approach is preferable here. It would avoid this loop here, and the logic might be more readable, but it likely has weirder edge cases (e.g. if "example.com" is blocked, we should prevent issuance for "foo.example.com" but not for "fooexample.com"). I'd be happy with either; just wanted to put this idea out there in case you liked it.

Reply:
I wrote it exactly like that till my head made me think about the domain prefix being problematic as you said. I was thinking about ways around it by adding more dots etc but always ended up with even more complex issues.